drivers/lis3mdl/lis3mdl.c

```diff
@@ -0,0 +1,216 @@
+{
+    int res;
+    char tmp;
+
+    i2c_acquire(dev->i2c);
+    DEBUG("LIS3MDL: Retrieving magnetometer data ... ");
+
+    /* enable block data update */
+    res = i2c_write_reg(dev->i2c, dev->mag_address,
+                         LIS3MDL_CTRL_REG5, LIS3MDL_MASK_REG5_BDU);
+
+    /* retrieve data from x-axis */
+    res += i2c_read_reg(dev->i2c, dev->mag_address,
+                        LIS3MDL_OUT_X_L_REG, &tmp);
+    data->x_axis = tmp;
+    res += i2c_read_reg(dev->i2c, dev->mag_address,
+                        LIS3MDL_OUT_X_H_REG, &tmp);
+    data->x_axis = tmp<<8; /* shift the 8 msb to left */
+
+    /* retrieve data from y-axis */
+    res += i2c_read_reg(dev->i2c, dev->mag_address,
+                        LIS3MDL_OUT_Y_L_REG, &tmp);
+    data->y_axis = tmp;
+    res += i2c_read_reg(dev->i2c, dev->mag_address,
+                        LIS3MDL_OUT_Y_H_REG, &tmp);
+    data->y_axis = tmp<<8;
+
+    /* retrieve data from z-axis */
+    res += i2c_read_reg(dev->i2c, dev->mag_address,
+                        LIS3MDL_OUT_Z_L_REG, &tmp);
+    data->z_axis = tmp;
+    res += i2c_read_reg(dev->i2c, dev->mag_address,
+                        LIS3MDL_OUT_Z_H_REG, &tmp);
+    data->z_axis = tmp<<8;
+
+    /* disable block data update */
+    res += i2c_write_reg(dev->i2c, dev->mag_address,
+                         LIS3MDL_CTRL_REG5, LIS3MDL_MASK_REG5_BDU_OFF);
+
+    i2c_release(dev->i2c);
+
+    if (res != 8)
+    {
+        DEBUG("[ FAIL ]\n");
+        return -1;
+    }
+
+    DEBUG("[ OK ]\n");
+    return 0;
+}
+
+int lis3mdl_read_temp(lis3mdl_t *dev, uint16 *value)
+{
+    int res;
+    char tmp;
+    uint16_t tmp_value;
+
+    i2c_acquire(dev->i2c);
+    DEBUG("LIS3MDL: Retrieving temperature data ... ");
+
+    /* enable block data update */
+    res = i2c_write_reg(dev->i2c, dev->mag_address,
+                         LIS3MDL_CTRL_REG5, LIS3MDL_MASK_REG5_BDU);
+
+    /* retrieve temperature data */
+    res += i2c_read_reg(dev->i2c, dev->mag_address,
+                        LIS3MDL_TEMP_OUT_L_REG, &tmp);
+    tmp_value = tmp;
+    res += i2c_read_reg(dev->i2c, dev->mag_address,
+                        LIS3MDL_TEMP_OUT_H_REG, &tmp);
+    tmp_value += tmp<<8;
+
+    /* disable block data update */
+    res += i2c_write_reg(dev->i2c, dev->mag_address,
+                         LIS3MDL_CTRL_REG5, LIS3MDL_MASK_REG5_BDU_OFF);
+
+    i2c_release(dev->i2c);
+
+    if (res != 4)
+    {
+        DEBUG("[ FAIL ]\n");
+        return -1;
+    }
+
+    *value = tmp_value;
+
+    DEBUG("[ OK ]\n");
+    return 0;
+}
+
+int lis3mdl_enable(lis3mdl_t *dev)
+{
+    int res;
+
+    i2c_acquire(dev->i2c);
+    DEBUG("LIS3MDL: Enable device ... ");
+
+    /* disable power down mode */
+    res = i2c_write_reg(dev->i2c, dev->mag_address,
+                        LIS3MDL_CTRL_REG3, LIS3MDL_MASK_REG3_LOW_POWER_DI);
+
+    /* enable temperature sensor */
+    res += i2c_write_reg(dev->i2c, dev->mag_address,
```

Comment:
@PeterKietzmann Check Page 24 - 27 if you want a quick view:

http://www.st.com/st-web-ui/static/active/en/resource/technical/document/datasheet/DM00075867.pdf